pkg/cloudfabric-controller/controllerframework/controller_framework_utils.go

```diff
@@ -54,53 +51,37 @@ func SortControllerInstancesByKeyAndConvertToLocal(controllerInstanceMap map[str
 	return sortedControllerInstancesLocal
 }
 
-// The following are test util. Put here so that can be used in other packages' tests
-
-var alwaysReady = func() bool { return true }
-var notifyTimes int
-
-func mockNotifyHander(controllerInstance *v1.ControllerInstance) {
-	notifyTimes++
-	return
-}
-func mockCheckInstanceHander() {
-	return
-}
-
-func CreateTestControllerInstanceManager(stopCh chan struct{}) (*ControllerInstanceManager, informers.SharedInformerFactory) {
-	client := fake.NewSimpleClientset()
-	informers := informers.NewSharedInformerFactory(client, 0)
-
-	cim := NewControllerInstanceManager(informers.Core().V1().ControllerInstances(), client, nil)
-	go cim.Run(stopCh)
-
-	cim.controllerListerSynced = alwaysReady
-	cim.notifyHandler = mockNotifyHander
-	checkInstanceHandler = mockCheckInstanceHander
-	return GetControllerInstanceManager(), informers
-}
+// Generate controllerKey for new controller instance. It is to find and split a scope for new controller instance.
+// Scope Splitting Principles:
+// 1. We always find existing scope with biggest size, and split it.
+// 2. If there are more than one scope at the biggest size, we chose the one with most ongoing work load, and split it.
+// 3. If both existing scope size and ongoing work are even, we choose first scope and split it.
+func GenerateKey(c *ControllerBase) int64 {
```

Comment:
Why is this method moved out of ControllerBase?

Reply:
Will need to check whether ControllerBase is nil

Reply:
Next iteration will be much more code and completed, so I think it's better to separate the logic out to another file.

Reply:
As long as we can keep a single function fit into a page, it should be fine.

Reply:
I believe next iteration I couldn't make it a single function. 
Current code "controller.controllerKey = controller.generateKey()", logically I feel it's weird. I think it would be proper to change to "controller.controllerKey = utils.GenerateKey()". And I will separate the burden to utils.

Reply:
Updated